README.md
# react-input-position

react-input-position is a component that decorates it's children with mouse/touch position tracking, a status toggle fired by click/gesture events, and more. Assisting in the creation of UI features that require detailed information about user touch/mouse input.

The status toggle feature can be triggered by Click, Double Click, Hover, Mouse Down, Tap, Double Tap, Touch, and Long Touch gestures.

Includes an item tracking feature which provides coordinates that can be moved by user input. This position can be limited through min/max settings or even limited based on the size of a linked element anywhere in the hierarchy of children. Useful for creating some image components such as magnifiers.

[![npm version](https://img.shields.io/npm/v/react-input-position.svg?style=flat)](https://npmjs.org/package/react-input-position "View this project on npm")
[![CircleCI Status](https://circleci.com/gh/AdamRisberg/react-input-position.svg?style=shield&circle-token=:circle-token)](https://circleci.com/gh/AdamRisberg/react-input-position)
[![npm downloads](https://img.shields.io/npm/dm/react-input-position.svg?style=flat-square)](https://www.npmjs.com/package/react-input-position)
[![MIT license](https://img.shields.io/badge/license-MIT-brightgreen.svg)](http://opensource.org/licenses/MIT)

## Demo

Visit the [react-input-position demo site](https://adamrisberg.github.io/react-input-position).

## Installation

```sh
npm install --save react-input-position
```

## Basic Usage

```JSX
import ReactInputPosition from "react-input-position";
...

<ReactInputPosition>
  <YourComponentOne />
  <YourComponentTwo />
</ReactInputPosition>
```

Importing and using activation method constants:

```JSX
import ReactInputPosition, {
  MOUSE_ACTIVATION,
  TOUCH_ACTIVATION
} from "react-input-position";
...

<ReactInputPoisiton
  mouseActivationMethod={MOUSE_ACTIVATION.DOUBLE_CLICK}
  touchActivationMethod={TOUCH_ACTIVATION.DOUBLE_TAP}
>
</ReactInputPosition>
```

Child components will receive the following props (although some may be inactive depending on the options you select):

```
{
  active: boolean,
  activePosition: {
    x: number,
    y: number
  },
  prevActivePosition: {
    x: number,
    y: number
  },
  passivePosition: {
    x: number,
    y: number
  },
  elementDimensions: {
    width: number,
    height: number
  },
  elementOffset: {
    left: number,
    top: number
  },
  itemPosition: {
    x: number,
    y: number
  },
  itemDimensions: {
    width: number,
    height: number
  },
  itemRef: React Ref,
  onLoadRefresh: function
}
```

**active:** The component's current state (active/inactive).

**activePosition:** Current position of mouse or touch event. Only updated while the component is in an active state.

**prevActivePosition:** Previous position of mouse of touch event. Helpful for calculating movement between updates. Only updated while the component is in an active state. Disabled by default.

**passivePosition:** Current position of mouse or touch event. Updated even when the component is in an inactive state. Disabled by default.

**elementDimensions:** Current dimensions of the component's wrapping div.

**elementOffset:** Current left and top offset of the component's wrapping div.

**itemDimensions:** Current dimensions of the tracked item. Disabled by default.

**itemRef:** Used when enabling the item position feature. Pass this to the element you wish to track. Example: `<div ref={props.itemRef}>Content</div>`.

**onLoadRefresh:** If slow loading elements are causing your initial stats to be inaccurate, you can call this function after they are loaded to reset and recalculate. For example, in an image's onload event.

## Props API

All props are optional.

**mouseActivationMethod:** Sets the mouse action necessary to toggle the component's active state. Accepted Values: "click", "doubleClick", "hover", "mouseDown". Type: string, Default: click.

**touchActivationMethod:** Sets the touch gesture necessary to toggle the component's active state. Accepted Values: "tap", "doubleTap", "touch", "longTouch". Type: string, Default: tap.

**tapDurationInMs:** Sets the maximum length of touch events in order to be considered tap gestures. Type: number, Default: 180.

**doubleTapDurationInMs:** Sets the minimum length of time in which two tap gestures must be performed in order to be considered a double tap gesture. Type: number, Default: 400.

**longTouchDurationInMs:** Sets the minimum length of touch events in order to be consider a long touch gesture. Type: number, Default: 500.

**longTouchMoveLimit:** Sets the maximum movement allowed during a long touch gesture. Type: number, Default: 5.

**clickMoveLimit:** Sets the maximum movement allowed during a mouse click. Helps to differentiate between a drag and click. Type: number, Default: 5.

**minUpdateSpeedInMs:** Throttles the mouse/touch position updates. For example, setting this to 10 will force a 10ms delay between position updates. A minimum of 10ms (since it's asynchronous). This won't actually cause updates every 10ms, it simply adds a delay to the natural rate. Note: Setting this to zero can create performance issues in IE. Raising this number too high will cause animations tied to positions to become choppy. Type: number, Default: 1.

**trackPassivePosition:** Provides mouse position regardless of active/inactive state. Type: boolean, Default: false.

**trackPreviousPosition:** Provides previous mouse/touch position. Useful for calculating movement between updates. Type: Boolean, Default: false.

**trackItemPosition:** Activates the item position functionality. Type: boolean, Default: false.

**centerItemOnLoad:** Centers the item position on component mount. Requires limits to be set for both min/max of both axis or limit by size. Type: boolean, Default: false.

**centerItemOnActivate:** Centers the item position whenever the component switches to an active state. Requires limits to be set for both min/max of both axis or limit by size. Type: boolean, Default: false.

**centerItemOnActivatePos:** Centers the item based on the current mouse/touch position whenever the component switches to an active state. Achieved by scaling the input position based on the sizes of the component and item. Useful for image zoom style components where you want to display a larger version of an image, centered on the spot where the user clicked/tapped. Requires use of the itemRef prop (see "Item Position Feature"). Type: boolean, Default: false.

**alignItemOnActivePos:** Links and aligns the item position with the active position. Achieved by scaling the active position based on the sizes of the component and item. Useful for creating components like a magnifying glass where a larger version of an image needs to stay aligned with a smaller version based on the current mouse/touch position. Requires use of the itemRef prop (see "Item Position Feature"). Type: boolean, Default: false.

**itemMovementMultiplier:** Multiplier to adjust the item's movement. Can also be used to reverse item movement by using a negative number. Type: number, Default: 1.

**linkItemToActive:** Links the item position to the active position. Type: boolean, Default: false.

**itemPositionMinX:** Sets the minimum value for the x axis of the item position. Type: number, Default: undefined.

**itemPositionMaxX:** Sets the maximum value for the x axis of the item position. Type: number, Default: undefined.

**itemPositionMinY:** Sets the minimum value for the y axis of the item position. Type: number, Default: undefined.

**itemPositionMaxY:** Sets the maximum value for the y axis of the item position. Type: number, Default: undefined.

**itemPositionLimitBySize:** Limits the item's position based on the size of the item. Assumes that the item is larger than the component's size. For example, moving around a large image while looking through a small viewing window. Requires use of the itemRef prop (see "Item Position Feature"). Type: boolean, Default: false.

**itemPositionLimitInternal:** Changes the behavior of itemPositionLimitBySize to assume that the item is smaller than the component's size. Stops the item from moving outside the bounds of the component. For example, moving around a small image within a large viewing window. Type: boolean, Default: false.

**mouseDownAllowOutside:** Allows component to stay active and continue tracking positions while mouse is outside of the tracked area. Only applies to the "mouseDown" activation method. Type: boolean, Default: false.

**className:** Passed to the className of the component's parent div. Type: string, Default: "".

**style:** Passed to the style of the component's parent div. Type: object, Default: {}.

**cursorStyle:** Sets the cursor style while the component is inactive. Accepts standard CSS cursor values. Type: string, Default: crosshair.

**cursorStyleActive:** Sets the cursor style while the component is active. Accepts standard CSS cursor values. Type: string, Default: Uses cursorStyle if not set.

**onActivate:** Callback function to be executed when component becomes active. Type: function.

**onDeactivate:** Callback function to be executed when component becomes inactive. Type: function.

## Inversion of Control Props

All children passed to the component expand the tracked interaction area (unless they have absolute positioning, for example). Because of this, sometimes, you may want to access the mouse/touch information in a sibling component. The following props can be used together in order to lift state to a parent component. All funcionality will remain the same. See the example below the prop descriptions.

**onUpdate:** Used to pass a callback function which gets called on every state change with an object representing the new state. When used in combination with the overrideState prop, this callback will be called in place of internal setState with an object containing proposed state changes. Type: function

**overrideState:** Used to pass a state object which will be used instead of internal state. Since this causes the component to surrendor control of state, it must be used in combination with the onUpdate callback. Type: object

```JSX
import ReactInputPosition, {
  MOUSE_ACTIVATION,
  TOUCH_ACTIVATION,
  defaultState
} from "react-input-position";

class ParentComponent extends React.Component {
  state = defaultState;

  onUpdate = (stateChanges) => {
    this.setState(stateChanges);
  };

  render() {
    return (
      <div>
        <ReactInputPoisiton
          mouseActivationMethod={MOUSE_ACTIVATION.DOUBLE_CLICK}
          touchActivationMethod={TOUCH_ACTIVATION.DOUBLE_TAP}
          overrideState={this.state}
          onUpdate={this.onUpdate}
        >
          <DecoratedChild />
        </ReactInputPosition>
        <SiblingComponent {...this.state} >
      </div>
    );
  }
}
```

## Item Position Feature

To use this feature, set the component's trackItemPosition prop to true. If you intend to use the itemPositionLimitBySize feature, pass the itemRef prop to the element you wish to enable the feature on. Example: `<img ref={props.itemRef} src="./example.jpg" />`.

Set the item element's position to absolute and use the itemPosition prop to position the element. Set the item position limits to control how far the item can be moved.

The item element will now move based on the input of the user while the component is in an active state. By default, this position will not reset for the life of the component.

You can adjust the behavior of this feature further using the many API props provided.

## Style Notes

Some styling passed to the component may be overwritten if they conflict with styles required for correct behavior.

By default, the width of the component will adjust to the size of the child components. You can change this behavior by passing in your own styling using the className and/or style props.

## Example Project

```sh
git clone https://github.com/adamrisberg/react-input-position.git
cd react-input-position
npm install
npm start
```


CHANGELOG.md
# Changelog

## 1.3.2 (November 6th, 2020)

- Fix undefined itemDimensions error.

## 1.3.1 (February 6th, 2020)

- Added onActivate and onDeactivate props. Allows passing callbacks to be executed when component becomes active or inactive, respectively.

## 1.3.0 (January 28th, 2020)

- Add mouseDownAllowOutside prop. Allows component to stay active and continue tracking positions while mouse is outside of the tracked area. Only applies to the "mouseDown" activation method.

## 1.2.1 (January 28th, 2020)

- Fixed item position tracking while using mouse down activation method

## 1.2.0 (September 26th, 2019)

- Added inversion of control props (onUpdate & overrideState)
- Fix touchmove resetting position issue

## 1.1.7 (September 1st, 2019)

- Fix non-passive touch event warnings

## 1.1.6 (August 8th, 2019)

- Removed unused properties in mouse and touch event handlers
- Removed unnecessary rounding in alignItemOnPosition
- Added tests for index, utils, and event handlers
- Added CircleCi
- Added CircleCi and downloads badges to README

## 1.1.5 (May 16th, 2019)

- Update README

## 1.1.4 (May 15th, 2019)

- Updated github links

## 1.1.3 (March 14th, 2019)

- Fixed bug when using negative item position limits.
- Added CHANGELOG.

## 1.1.2 (March 12th, 2019)

- Refactored code to improve readability and performance.

## 1.1.1 (January 14th, 2019)

- Fixed typos in README.

## 1.1.0 (January 13th, 2019)

- Added Align Item on Active Position feature.
- Various code cleanup.

## 1.0.1 (January 10th, 2019)

- Updated webpack-dev-server to remove vulnerability.
- Updated README.
- Added license.

## 1.0.0 (January 10th, 2019)

- Initial public release.


src/index.js
import React, { Component } from "react";
import PropTypes from "prop-types";
import mouseActivation from "./mouseActivation";
import touchActivation from "./touchActivation";
import { MOUSE_ACTIVATION, TOUCH_ACTIVATION } from "./constants";
import utils from "./utils";

const defaultState = {
  active: false,
  activePosition: { x: 0, y: 0 },
  prevActivePosition: { x: 0, y: 0 },
  passivePosition: { x: 0, y: 0 },
  elementDimensions: { width: 0, height: 0 },
  elementOffset: { left: 0, top: 0 },
  itemPosition: { x: 0, y: 0 },
  itemDimensions: { width: 0, height: 0 }
};

class ReactInputPosition extends Component {
  state = defaultState;

  containerRef = React.createRef();
  itemRef = React.createRef();
  mouseDown = false;
  touched = false;
  justTouched = false;
  tapped = false;
  tapTimer = null;
  tapTimedOut = false;
  doubleTapTimer = null;
  doubleTapTimedOut = false;
  longTouchTimer = null;
  longTouchTimedOut = false;
  refresh = true;

  static propTypes = {
    mouseActivationMethod: PropTypes.oneOf([
      MOUSE_ACTIVATION.CLICK,
      MOUSE_ACTIVATION.DOUBLE_CLICK,
      MOUSE_ACTIVATION.HOVER,
      MOUSE_ACTIVATION.MOUSE_DOWN
    ]).isRequired,
    touchActivationMethod: PropTypes.oneOf([
      TOUCH_ACTIVATION.DOUBLE_TAP,
      TOUCH_ACTIVATION.LONG_TOUCH,
      TOUCH_ACTIVATION.TAP,
      TOUCH_ACTIVATION.TOUCH
    ]).isRequired,
    tapDurationInMs: PropTypes.number,
    doubleTapDurationInMs: PropTypes.number,
    longTouchDurationInMs: PropTypes.number,
    longTouchMoveLimit: PropTypes.number,
    clickMoveLimit: PropTypes.number,
    itemPositionMinX: PropTypes.number,
    itemPositionMaxX: PropTypes.number,
    itemPositionMinY: PropTypes.number,
    itemPositionMaxY: PropTypes.number,
    itemPositionLimitBySize: PropTypes.bool,
    itemPositionLimitInternal: PropTypes.bool,
    linkItemToActive: PropTypes.bool,
    className: PropTypes.string,
    style: PropTypes.object,
    minUpdateSpeedInMs: PropTypes.number,
    trackPassivePosition: PropTypes.bool,
    trackItemPosition: PropTypes.bool,
    trackPreviousPosition: PropTypes.bool,
    centerItemOnActivate: PropTypes.bool,
    centerItemOnActivatePos: PropTypes.bool,
    centerItemOnLoad: PropTypes.bool,
    alignItemOnActivePos: PropTypes.bool,
    itemMovementMultiplier: PropTypes.number,
    cursorStyle: PropTypes.string,
    cursorStyleActive: PropTypes.string,
    onUpdate: PropTypes.func,
    overrideState: PropTypes.object,
    mouseDownAllowOutside: PropTypes.bool,
    onActivate: PropTypes.func,
    onDeactivate: PropTypes.func
  };

  static defaultProps = {
    tapDurationInMs: 180,
    doubleTapDurationInMs: 400,
    longTouchDurationInMs: 500,
    longTouchMoveLimit: 5,
    clickMoveLimit: 5,
    style: {},
    minUpdateSpeedInMs: 1,
    itemMovementMultiplier: 1,
    cursorStyle: "crosshair",
    mouseActivationMethod: MOUSE_ACTIVATION.CLICK,
    touchActivationMethod: TOUCH_ACTIVATION.TAP,
    mouseDownAllowOutside: false
  };

  componentDidMount() {
    this.init();
    this.refreshPosition();
  }

  componentWillUnmount() {
    this.removeMouseEventListeners();
    this.removeTouchEventListeners();
    this.removeOtherEventListeners();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.mouseActivationMethod !== this.props.mouseActivationMethod) {
      this.removeMouseEventListeners();
      this.setMouseInteractionMethods();
      this.addMouseEventListeners();
    }
    if (prevProps.touchActivationMethod !== this.props.touchActivationMethod) {
      this.removeTouchEventListeners();
      this.setTouchInteractionMethods();
      this.addTouchEventListeners();
    }
  }

  init() {
    this.checkPassiveEventSupport();
    this.setInputInteractionMethods();
    this.addMouseEventListeners();
    this.addTouchEventListeners();
    this.addOtherEventListeners();
  }

  checkPassiveEventSupport() {
    this.supportsPassive = false;
    try {
      const options = Object.defineProperty({}, "passive", {
        get: () => (this.supportsPassive = true)
      });
      window.addEventListener("testPassive", null, options);
      window.removeEventListener("testPassive", null, options);
    } catch (e) {}
  }

  updateState(changes, cb) {
    const { onUpdate } = this.props;

    let activationCallback;
    if (changes.hasOwnProperty("active")) {
      if (changes.active) {
        activationCallback = this.props.onActivate;
      } else {
        activationCallback = this.props.onDeactivate;
      }
    }

    if (this.props.overrideState) {
      onUpdate && onUpdate(changes);
      activationCallback && activationCallback();
      cb && cb.call(this);
      return;
    }

    this.setState(
      () => changes,
      () => {
        cb && cb.call(this);
        activationCallback && activationCallback();
        onUpdate && onUpdate(this.getStateClone());
      }
    );
  }

  getState() {
    if (this.props.overrideState) {
      return this.props.overrideState;
    } else {
      return this.state;
    }
  }

  getStateClone() {
    const state = this.getState();
    const clonedState = {};

    for (let key in state) {
      if (typeof state[key] === "object") {
        clonedState[key] = { ...state[key] };
      } else {
        clonedState[key] = state[key];
      }
    }

    return clonedState;
  }

  onLoadRefresh = () => {
    this.refreshPosition();
  };

  refreshPosition() {
    const { trackItemPosition, centerItemOnLoad } = this.props;

    this.setPosition(
      { x: 0, y: 0 },
      trackItemPosition,
      false,
      centerItemOnLoad
    );
  }

  setInputInteractionMethods() {
    this.setMouseInteractionMethods();
    this.setTouchInteractionMethods();
  }

  setMouseInteractionMethods() {
    const mouseInteractionMethods =
      mouseActivation[this.props.mouseActivationMethod];
    this.mouseHandlers = [];

    for (let key in mouseInteractionMethods) {
      this.mouseHandlers.push({
        event: key.toLowerCase(),
        handler: mouseInteractionMethods[key].bind(this)
      });
    }
  }

  setTouchInteractionMethods() {
    const touchInteractionMethods =
      touchActivation[this.props.touchActivationMethod];
    this.touchHandlers = [];

    for (let key in touchInteractionMethods) {
      this.touchHandlers.push({
        event: key.toLowerCase(),
        handler: touchInteractionMethods[key].bind(this)
      });
    }
  }

  handleResize = () => {
    this.refreshPosition();
  };

  setPosition(position, updateItemPosition, activate, centerItem) {
    if (this.props.minUpdateSpeedInMs && !this.refresh) return;
    this.refresh = false;

    const {
      left,
      top,
      width,
      height
    } = this.containerRef.current.getBoundingClientRect();

    const {
      trackItemPosition,
      trackPassivePosition,
      trackPreviousPosition,
      centerItemOnActivate,
      centerItemOnActivatePos,
      linkItemToActive,
      itemMovementMultiplier,
      alignItemOnActivePos,
      itemPositionMinX,
      itemPositionMaxX,
      itemPositionMinY,
      itemPositionMaxY,
      itemPositionLimitBySize,
      itemPositionLimitInternal
    } = this.props;

    const { activePosition, itemPosition } = this.getState();

    // Set container div info and active position
    const stateUpdate = {
      elementDimensions: { width, height },
      elementOffset: { left, top },
      activePosition: {
        x: Math.min(Math.max(0, position.x - left), width),
        y: Math.min(Math.max(0, position.y - top), height)
      }
    };

    // Activate if necessary
    if (activate) stateUpdate.active = true;

    // Set item dimensions
    if (this.itemRef.current) {
      const itemSize = this.itemRef.current.getBoundingClientRect();

      stateUpdate.itemDimensions = {
        width: itemSize.width,
        height: itemSize.height
      };
    }

    // Set previous active position
    if (trackPreviousPosition || trackItemPosition) {
      stateUpdate.prevActivePosition = {
        x: activePosition.x,
        y: activePosition.y
      };
    }

    // Set passive position
    if (trackPassivePosition) {
      stateUpdate.passivePosition = {
        x: position.x - left,
        y: position.y - top
      };
    }

    // Create adjusted limits
    const limits = utils.createAdjustedLimits(
      itemPositionMinX,
      itemPositionMaxX,
      itemPositionMinY,
      itemPositionMaxY,
      stateUpdate.elementDimensions,
      stateUpdate.itemDimensions,
      itemPositionLimitBySize,
      itemPositionLimitInternal
    );

    // Center item
    if (centerItem || (activate && centerItemOnActivate)) {
      const centerX = (limits.maxX + limits.minX) / 2;
      const centerY = (limits.maxY + limits.minY) / 2;

      stateUpdate.itemPosition = {
        x: centerX || 0,
        y: centerY || 0
      };

      return this.updateState(stateUpdate, this.startRefreshTimer);
    }

    let shouldLimitItem = true;

    // Set item position
    if (trackItemPosition && linkItemToActive) {
      stateUpdate.itemPosition = { ...stateUpdate.activePosition };
    } else if (trackItemPosition && alignItemOnActivePos) {
      stateUpdate.itemPosition = utils.alignItemOnPosition(
        stateUpdate.elementDimensions,
        stateUpdate.itemDimensions,
        stateUpdate.activePosition
      );
    } else if (trackItemPosition && activate && centerItemOnActivatePos) {
      stateUpdate.itemPosition = utils.centerItemOnPosition(
        stateUpdate.elementDimensions,
        stateUpdate.itemDimensions,
        stateUpdate.activePosition
      );
    } else if (trackItemPosition && updateItemPosition) {
      stateUpdate.itemPosition = utils.calculateItemPosition(
        itemPosition,
        stateUpdate.prevActivePosition,
        stateUpdate.activePosition,
        itemMovementMultiplier
      );
    } else {
      shouldLimitItem = false;
    }

    // Apply position limits
    if (shouldLimitItem) {
      stateUpdate.itemPosition = utils.limitPosition(
        limits.minX,
        limits.maxX,
        limits.minY,
        limits.maxY,
        stateUpdate.itemPosition
      );
    }

    this.updateState(stateUpdate, this.startRefreshTimer);
  }

  setPassivePosition(position) {
    if (!this.props.trackPassivePosition) return;

    const { left, top } = this.containerRef.current.getBoundingClientRect();

    this.updateState({
      passivePosition: {
        x: position.x - left,
        y: position.y - top
      }
    });
  }

  toggleActive(position = { x: 0, y: 0 }) {
    if (!this.getState().active) {
      this.activate(position);
    } else {
      this.deactivate();
    }
  }

  activate(position = { x: 0, y: 0 }) {
    this.setPosition(position, false, true);
  }

  deactivate() {
    this.updateState({ active: false });
  }

  startRefreshTimer() {
    if (!this.props.minUpdateSpeedInMs) return;

    setTimeout(() => {
      this.refresh = true;
    }, this.props.minUpdateSpeedInMs);
  }

  startTapTimer() {
    this.tapTimer = setTimeout(() => {
      this.tapTimedOut = true;
    }, this.props.tapDurationInMs);
  }

  startDoubleTapTimer() {
    this.doubleTapTimer = setTimeout(() => {
      this.doubleTapTimedOut = true;
    }, this.props.doubleTapDurationInMs);
  }

  startLongTouchTimer(e) {
    this.longTouchTimer = setTimeout(() => {
      if (this.touched) {
        this.toggleActive(e);
      }
    }, this.props.longTouchDurationInMs);
  }

  addMouseEventListeners() {
    this.mouseHandlers.forEach(mouse => {
      this.containerRef.current.addEventListener(mouse.event, mouse.handler);
    });
  }

  addTouchEventListeners() {
    this.touchHandlers.forEach(touch => {
      this.containerRef.current.addEventListener(
        touch.event,
        touch.handler,
        this.supportsPassive ? { passive: false } : false
      );
    });
  }

  removeMouseEventListeners() {
    this.mouseHandlers.forEach(mouse => {
      this.containerRef.current.removeEventListener(mouse.event, mouse.handler);
    });
  }

  removeTouchEventListeners() {
    this.touchHandlers.forEach(touch => {
      this.containerRef.current.removeEventListener(
        touch.event,
        touch.handler,
        this.supportsPassive ? { passive: false } : false
      );
    });
  }

  addOtherEventListeners() {
    window.addEventListener("resize", this.handleResize);
    window.addEventListener("load", this.onLoadRefresh);
  }

  removeOtherEventListeners() {
    window.removeEventListener("resize", this.handleResize);
    window.removeEventListener("load", this.onLoadRefresh);
  }

  render() {
    const {
      style,
      className,
      children,
      cursorStyle,
      cursorStyleActive
    } = this.props;
    const { active } = this.getState();

    const combinedStyle = {
      ...style,
      WebkitUserSelect: "none",
      MozUserSelect: "none",
      msUserSelect: "none",
      userSelect: "none",
      cursor: active ? cursorStyleActive || cursorStyle : cursorStyle
    };

    return (
      <div style={combinedStyle} className={className} ref={this.containerRef}>
        {utils.decorateChildren(children, {
          ...this.getState(),
          itemRef: this.itemRef,
          onLoadRefresh: this.onLoadRefresh
        })}
      </div>
    );
  }
}

export { MOUSE_ACTIVATION, TOUCH_ACTIVATION, defaultState };
export default ReactInputPosition;
